File: cineforge-h5/src/assetNormalization.ts
import type {
  AssetNormalizationRelations,
  AssetNormalizationGlobalReviewItem,
  AssetNormalizationReviewItem,
  AssetNormalizationV1,
  AssetRelationTarget,
  AssetTypeV1,
  BreakdownDraftAsset,
  BreakdownDraftData,
  BreakdownRead,
} from "./types.ts";

export const ASSET_NORMALIZATION_VERSION = "AssetNormalization.v1" as const;

type UnknownRecord = Record<string, unknown>;

export function assertAssetNormalizationV1(value: unknown): asserts value is AssetNormalizationV1 {
  const asset = requiredRecord(value, "asset");
  exactValue(asset.normalization_version, ASSET_NORMALIZATION_VERSION, "asset.normalization_version");
  const clientAssetKey = nonEmptyString(asset.client_asset_key, "asset.client_asset_key");
  const assetCode = nonEmptyString(asset.asset_code, "asset.asset_code");
  if (!/^[A-Z0-9]+-[A-Z][A-Z0-9]*$/i.test(assetCode)) {
    invalid("asset.asset_code", "a complete project-prefixed asset code");
  }
  const displayCode = nonEmptyString(asset.display_code, "asset.display_code");
  const displayLabel = nonEmptyString(asset.display_label, "asset.display_label");
  enumValue(asset.code_state, ["candidate", "formal"] as const, "asset.code_state");
  const assetType = enumValue(asset.asset_type, ["character", "scene", "prop"] as const, "asset.asset_type");
  const name = nonEmptyString(asset.name, "asset.name");
  const typePrefix = assetType === "character" ? "R" : assetType === "scene" ? "SC" : "P";
  if (!displayCode.startsWith(typePrefix)) invalid("asset.display_code", `a ${typePrefix} asset code`);
  if (!assetCode.endsWith(`-${displayCode}`)) invalid("asset.asset_code", "the same identity as display_code");
  if (displayLabel !== `${displayCode}-${name}`) invalid("asset.display_label", "display_code joined with name");
  nullableString(asset.description, "asset.description");
  enumValue(asset.priority, ["S", "A", "B", "C"] as const, "asset.priority");
  assertAttributes(asset.attributes, assetType, "asset.attributes");
  assertRelations(asset.relations, "asset.relations");
  requiredArray(asset.review_items, "asset.review_items").forEach((item, index) => {
    assertReviewItem(item, clientAssetKey, `asset.review_items[${index}]`);
  });
  assertSource(asset.source, "asset.source");
  requiredArray(asset.repairs, "asset.repairs").forEach((item, index) => {
    assertRepair(item, `asset.repairs[${index}]`);
  });
}

export function parseAssetNormalizationV1(value: unknown): AssetNormalizationV1 {
  assertAssetNormalizationV1(value);
  return value;
}

export function parseAssetNormalizations(value: unknown): AssetNormalizationV1[] {
  const clientKeys = new Set<string>();
  const assetCodes = new Set<string>();
  const displayCodes = new Set<string>();
  const assets = requiredArray(value, "assets").map((asset, index) => {
    try {
      const parsed = parseAssetNormalizationV1(asset);
      if (clientKeys.has(parsed.client_asset_key)) {
        invalid(`assets[${index}].client_asset_key`, "a unique client_asset_key");
      }
      if (assetCodes.has(parsed.asset_code)) invalid(`assets[${index}].asset_code`, "a unique asset_code");
      if (displayCodes.has(parsed.display_code)) invalid(`assets[${index}].display_code`, "a unique display_code");
      clientKeys.add(parsed.client_asset_key);
      assetCodes.add(parsed.asset_code);
      displayCodes.add(parsed.display_code);
      return parsed;
    } catch (error) {
      if (error instanceof TypeError) {
        throw new TypeError(`assets[${index}]: ${error.message}`);
      }
      throw error;
    }
  });
  const byKey = new Map(assets.map((asset) => [asset.client_asset_key, asset]));
  const reviewKeys = new Set<string>();
  assets.forEach((asset, assetIndex) => {
    asset.review_items.forEach((item, reviewIndex) => {
      const reviewKey = `${item.asset_key}\u0000${item.review_id}`;
      if (reviewKeys.has(reviewKey)) invalid(`assets[${assetIndex}].review_items[${reviewIndex}].review_id`, "unique within its asset");
      reviewKeys.add(reviewKey);
    });
    asset.relations.scene_links.forEach((link, linkIndex) => {
      assertClosedRelation(link, byKey, "scene", `assets[${assetIndex}].relations.scene_links[${linkIndex}]`);
    });
    if (asset.relations.owner_character) {
      assertClosedRelation(asset.relations.owner_character, byKey, "character", `assets[${assetIndex}].relations.owner_character`);
    }
  });
  return assets;
}

export function parseGlobalAssetReviewItems(value: unknown): AssetNormalizationGlobalReviewItem[] {
  const reviewIds = new Set<string>();
  return requiredArray(value, "global_review_items").map((item, index) => {
    const path = `global_review_items[${index}]`;
    const review = requiredRecord(item, path);
    exactKeys(review, [
      "review_id", "target_scope", "target_field", "issue", "suggestion", "status",
      "severity", "requirement_level", "source", "source_path", "human_input",
    ], path);
    const reviewId = nonEmptyString(review.review_id, `${path}.review_id`);
    if (reviewIds.has(reviewId)) invalid(`${path}.review_id`, "a unique review_id");
    reviewIds.add(reviewId);
    exactValue(review.target_scope, "inventory", `${path}.target_scope`);
    nonEmptyString(review.target_field, `${path}.target_field`);
    nonEmptyString(review.issue, `${path}.issue`);
    nullableString(review.suggestion, `${path}.suggestion`);
    nonEmptyString(review.status, `${path}.status`);
    enumValue(review.severity, ["warning", "error"] as const, `${path}.severity`);
    enumValue(review.requirement_level, ["required", "optional"] as const, `${path}.requirement_level`);
    nonEmptyString(review.source, `${path}.source`);
    nullableString(review.source_path, `${path}.source_path`);
    nullableString(review.human_input, `${path}.human_input`);
    return review as AssetNormalizationGlobalReviewItem;
  });
}

function assertClosedRelation(
  link: AssetRelationTarget,
  byKey: Map<string, AssetNormalizationV1>,
  expectedType: AssetTypeV1,
  path: string,
): void {
  const target = byKey.get(link.target_asset_key);
  if (!target) invalid(`${path}.target_asset_key`, "an asset in the same inventory");
  if (target.asset_type !== expectedType) invalid(`${path}.target_asset_key`, `a ${expectedType} asset`);
  if (
    link.target_asset_code !== target.asset_code
    || link.target_display_code !== target.display_code
    || link.target_display_label !== target.display_label
  ) {
    invalid(path, "the exact target asset identity");
  }
}

export function parseBreakdownRead(value: unknown): BreakdownRead {
  const breakdown = requiredRecord(value, "breakdown");
  exactKeys(breakdown, ["project_id", "version", "data_state", "view", "agent_run_id", "updated_at"], "breakdown");
  nonEmptyString(breakdown.project_id, "breakdown.project_id");
  if (!Number.isInteger(breakdown.version) || Number(breakdown.version) < 1) {
    invalid("breakdown.version", "a positive integer");
  }
  enumValue(breakdown.data_state, ["agent_raw", "normalized", "human_revision", "final"] as const, "breakdown.data_state");
  const view = requiredRecord(breakdown.view, "breakdown.view");
  exactValue(view.normalization_version, ASSET_NORMALIZATION_VERSION, "breakdown.view.normalization_version");
  parseAssetNormalizations(view.assets);
  parseGlobalAssetReviewItems(view.global_review_items);
  assertBreakdownReviewState(view.reading_review_state, "breakdown.view.reading_review_state");
  assertBreakdownReviewState(view.asset_review_state, "breakdown.view.asset_review_state");
  if (breakdown.agent_run_id !== null) nonEmptyString(breakdown.agent_run_id, "breakdown.agent_run_id");
  nonEmptyString(breakdown.updated_at, "breakdown.updated_at");
  return breakdown as BreakdownRead;
}

function assertBreakdownReviewState(value: unknown, path: string): void {
  const state = requiredRecord(value, path);
  onlyKeys(state, ["status", "confirmed_revision_id", "confirmation_mode"], path);
  enumValue(state.status, ["needs_review", "pending_confirmation", "confirmed"] as const, `${path}.status`);
  if (Object.prototype.hasOwnProperty.call(state, "confirmed_revision_id")) nullableString(state.confirmed_revision_id, `${path}.confirmed_revision_id`);
  if (Object.prototype.hasOwnProperty.call(state, "confirmation_mode")) nullableString(state.confirmation_mode, `${path}.confirmation_mode`);
}

export function assetByType(assets: readonly AssetNormalizationV1[], assetType: AssetTypeV1): AssetNormalizationV1[] {
  return assets.filter((asset) => asset.asset_type === assetType);
}

export function assetByClientKey(
  assets: readonly AssetNormalizationV1[],
  clientAssetKey: string,
): AssetNormalizationV1 | undefined {
  return assets.find((asset) => asset.client_asset_key === clientAssetKey);
}

export function relationsByClientAssetKey(
  assets: readonly AssetNormalizationV1[],
  clientAssetKey: string,
): AssetNormalizationRelations | undefined {
  return assetByClientKey(assets, clientAssetKey)?.relations;
}

export function reviewItemsByClientAssetKey(
  assets: readonly AssetNormalizationV1[],
  clientAssetKey: string,
): AssetNormalizationReviewItem[] | undefined {
  return assetByClientKey(assets, clientAssetKey)?.review_items;
}

export function assetDisplayCode(asset: AssetNormalizationV1): string {
  return asset.display_code;
}

export function assetDisplayLabel(asset: AssetNormalizationV1): string {
  return asset.display_label;
}

export function assetCodeIsFormal(asset: AssetNormalizationV1): boolean {
  return asset.code_state === "formal";
}

export function withEditedAssetInventory(draft: BreakdownDraftData, assets: BreakdownDraftAsset[]): BreakdownDraftData {
  return {
    ...draft,
    assets,
    assetReviewState: {
      ...draft.assetReviewState,
      status: "pending_confirmation",
    },
  };
}

export function withFreshAssetConfirmation(draft: BreakdownDraftData): BreakdownDraftData {
  return {
    ...draft,
    assetReviewState: {
      ...draft.assetReviewState,
      status: "confirmed",
      confirmed_revision_id: null,
      confirmation_mode: "human_confirmed",
    },
  };
}

export function breakdownPayload(draft: BreakdownDraftData, episodeCode = "EP01") {
  const revisionContext = draft.readingRevisionContext || {};
  const resolvedEpisodeCode = scopeString(revisionContext.episode_code) || episodeCode;
  // Only send fields the human actually edits in the workbench. Agent-derived,
  // display-only fields (readthrough, dialogue_script, coverage_checks, keyframe_plan,
  // video_plan, cross_references, delivery_checklist, costume/prompt summaries, etc.)
  // are NOT round-tripped — echoing them back re-persisted stale Agent output into the
  // human revision, violating the single-source version chain. The backend keeps its
  // own copies of those in the normalized layer.
  return {
    view: {
      normalization_version: ASSET_NORMALIZATION_VERSION,
      script_segments: draft.scriptSegments,
      storyboards: draft.storyboards.map((item, index) => ({
        ...item,
        order_num: item.order_num || index + 1,
        episode_code: resolvedEpisodeCode,
      })),
      assets: parseAssetNormalizations(draft.assets),
      global_review_items: parseGlobalAssetReviewItems(draft.globalAssetReviewItems),
      reading_report: draft.readingReport,
      reading_review_state: draft.readingReviewState,
      asset_review_state: draft.assetReviewState,
      notes: draft.notes,
      reading_revision_context: draft.readingRevisionContext,
    },
    episode_code: resolvedEpisodeCode,
    episode_id: scopeString(revisionContext.episode_id) || undefined,
    script_version_id: scopeString(revisionContext.script_version_id) || undefined,
    change_summary: ["前端人工修正保存"],
    source_label: "human_modified",
    data_state: "human_revision",
    expected_version: draft.version,
  };
}

function assertRelations(value: unknown, path: string): void {
  const relations = requiredRecord(value, path);
  exactKeys(relations, [
    "relationship_text", "scene_links", "owner_character", "storyboard_codes",
    "script_segment_codes", "evidence", "confidence",
  ], path);
  nullableString(relations.relationship_text, `${path}.relationship_text`);
  requiredArray(relations.scene_links, `${path}.scene_links`).forEach((item, index) => {
    assertRelationTarget(item, `${path}.scene_links[${index}]`);
  });
  if (relations.owner_character !== null) {
    assertRelationTarget(relations.owner_character, `${path}.owner_character`);
  }
  stringArray(relations.storyboard_codes, `${path}.storyboard_codes`);
  stringArray(relations.script_segment_codes, `${path}.script_segment_codes`);
  requiredArray(relations.evidence, `${path}.evidence`).forEach((item, index) => {
    assertEvidence(item, `${path}.evidence[${index}]`);
  });
  enumValue(relations.confidence, ["high", "medium", "low", "unknown"] as const, `${path}.confidence`);
}

function assertRelationTarget(value: unknown, path: string): asserts value is AssetRelationTarget {
  const target = requiredRecord(value, path);
  exactKeys(target, ["target_asset_key", "target_asset_code", "target_display_code", "target_display_label"], path);
  nonEmptyString(target.target_asset_key, `${path}.target_asset_key`);
  nonEmptyString(target.target_asset_code, `${path}.target_asset_code`);
  nonEmptyString(target.target_display_code, `${path}.target_display_code`);
  nonEmptyString(target.target_display_label, `${path}.target_display_label`);
}

function assertReviewItem(value: unknown, expectedAssetKey: string, path: string): void {
  const item = requiredRecord(value, path);
  exactKeys(item, [
    "review_id", "asset_key", "target_field", "issue", "suggestion", "status",
    "severity", "requirement_level", "source", "source_path", "human_input",
  ], path);
  nonEmptyString(item.review_id, `${path}.review_id`);
  exactValue(item.asset_key, expectedAssetKey, `${path}.asset_key`);
  nonEmptyString(item.target_field, `${path}.target_field`);
  nonEmptyString(item.issue, `${path}.issue`);
  nullableString(item.suggestion, `${path}.suggestion`);
  nonEmptyString(item.status, `${path}.status`);
  enumValue(item.severity, ["warning", "error"] as const, `${path}.severity`);
  enumValue(item.requirement_level, ["required", "optional"] as const, `${path}.requirement_level`);
  nonEmptyString(item.source, `${path}.source`);
  nullableString(item.source_path, `${path}.source_path`);
  nullableString(item.human_input, `${path}.human_input`);
}

function assertSource(value: unknown, path: string): void {
  const source = requiredRecord(value, path);
  exactKeys(source, ["source_kind", "agent_run_id", "skill_name", "raw_group", "raw_index"], path);
  enumValue(source.source_kind, ["agent", "human", "confirmed_inventory"] as const, `${path}.source_kind`);
  nullableString(source.agent_run_id, `${path}.agent_run_id`);
  nullableString(source.skill_name, `${path}.skill_name`);
  enumValue(source.raw_group, ["assets", "characters", "scenes", "props"] as const, `${path}.raw_group`);
  if (!Number.isInteger(source.raw_index) || Number(source.raw_index) < 0) {
    invalid(`${path}.raw_index`, "a non-negative integer");
  }
}

function assertRepair(value: unknown, path: string): void {
  const repair = requiredRecord(value, path);
  exactKeys(repair, ["field", "action", "before", "after", "reason"], path);
  stringValue(repair.field, `${path}.field`);
  stringValue(repair.action, `${path}.action`);
  requiredField(repair, "before", path);
  requiredField(repair, "after", path);
  stringValue(repair.reason, `${path}.reason`);
}

function assertAttributes(value: unknown, assetType: AssetTypeV1, path: string): void {
  const attributes = requiredRecord(value, path);
  if (assetType === "character") {
    exactKeys(attributes, [
      "character_type", "visual_presence", "role", "gender", "age", "appearance",
      "core_requirement", "personality", "background", "has_dialogue",
      "dialogue_evidence", "context_codes",
    ], path);
    for (const field of ["character_type", "role", "gender", "age", "appearance", "core_requirement", "personality", "background"]) {
      nullableString(attributes[field], `${path}.${field}`);
    }
    enumValue(attributes.visual_presence, ["on_screen", "voice_only", "mentioned_only"] as const, `${path}.visual_presence`);
    if (typeof attributes.has_dialogue !== "boolean") invalid(`${path}.has_dialogue`, "a boolean");
    stringArray(attributes.dialogue_evidence, `${path}.dialogue_evidence`);
    stringArray(attributes.context_codes, `${path}.context_codes`);
    return;
  }
  if (assetType === "scene") {
    exactKeys(attributes, [
      "context_code", "render_mode", "scene_no", "interior_exterior", "time", "location",
      "atmosphere", "narrative_function", "spatial_zones", "visual_goal", "key_props",
      "key_prop_codes", "continuity_risks", "episode_code",
    ], path);
    for (const field of ["context_code", "scene_no", "interior_exterior", "time", "location", "atmosphere", "narrative_function", "visual_goal", "episode_code"]) {
      nullableString(attributes[field], `${path}.${field}`);
    }
    if (attributes.render_mode !== null) {
      enumValue(attributes.render_mode, ["animated", "live_action"] as const, `${path}.render_mode`);
    }
    for (const field of ["spatial_zones", "key_props", "key_prop_codes", "continuity_risks"]) {
      stringArray(attributes[field], `${path}.${field}`);
    }
    return;
  }
  exactKeys(attributes, [
    "level", "prop_type", "appearance", "source", "status_change", "variant_states",
    "usage_scene", "scene_description", "context_codes",
  ], path);
  for (const field of ["level", "prop_type", "appearance", "source", "status_change", "usage_scene", "scene_description"]) {
    nullableString(attributes[field], `${path}.${field}`);
  }
  stringArray(attributes.variant_states, `${path}.variant_states`);
  stringArray(attributes.context_codes, `${path}.context_codes`);
}

function assertEvidence(value: unknown, path: string): void {
  const evidence = requiredRecord(value, path);
  exactKeys(evidence, ["target_code", "storyboard_code", "script_segment_code", "source_field", "excerpt"], path);
  for (const field of ["target_code", "storyboard_code", "script_segment_code", "source_field", "excerpt"]) {
    nullableString(evidence[field], `${path}.${field}`);
  }
}

function requiredRecord(value: unknown, path: string): UnknownRecord {
  if (!value || typeof value !== "object" || Array.isArray(value)) invalid(path, "an object");
  return value as UnknownRecord;
}

function scopeString(value: unknown): string {
  return value === null || value === undefined ? "" : String(value).trim();
}

function requiredArray(value: unknown, path: string): unknown[] {
  if (!Array.isArray(value)) invalid(path, "an array");
  return value;
}

function nonEmptyString(value: unknown, path: string): string {
  if (typeof value !== "string" || !value.trim()) invalid(path, "a non-empty string");
  return value;
}

function stringValue(value: unknown, path: string): string {
  if (typeof value !== "string") invalid(path, "a string");
  return value;
}

function nullableString(value: unknown, path: string): void {
  if (value !== null && typeof value !== "string") invalid(path, "a string or null");
}

function stringArray(value: unknown, path: string): void {
  requiredArray(value, path).forEach((item, index) => stringValue(item, `${path}[${index}]`));
}

function exactKeys(record: UnknownRecord, fields: readonly string[], path: string): void {
  const expected = new Set(fields);
  for (const field of fields) requiredField(record, field, path);
  const extra = Object.keys(record).find((field) => !expected.has(field));
  if (extra) invalid(`${path}.${extra}`, "no unknown fields");
}

function onlyKeys(record: UnknownRecord, fields: readonly string[], path: string): void {
  const expected = new Set(fields);
  const extra = Object.keys(record).find((field) => !expected.has(field));
  if (extra) invalid(`${path}.${extra}`, "no unknown fields");
}

function enumValue<const T extends readonly string[]>(value: unknown, allowed: T, path: string): T[number] {
  if (typeof value !== "string" || !allowed.includes(value)) invalid(path, allowed.join(" | "));
  return value as T[number];
}

function exactValue(value: unknown, expected: string, path: string): void {
  if (value !== expected) invalid(path, JSON.stringify(expected));
}

function requiredField(record: UnknownRecord, field: string, path: string): void {
  if (!Object.prototype.hasOwnProperty.call(record, field)) invalid(`${path}.${field}`, "a present field");
}

function invalid(path: string, expected: string): never {
  throw new TypeError(`Invalid AssetNormalization.v1 contract at ${path}; expected ${expected}`);
}

File: cineforge-server/agent/app/schemas/asset_normalization.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


NormalizationVersion = Literal["AssetNormalization.v1"]
AssetTypeV1 = Literal["character", "scene", "prop"]
AssetPriorityV1 = Literal["S", "A", "B", "C"]
AssetCodeStateV1 = Literal["candidate", "formal"]


class CharacterAttributesV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    character_type: str | None = None
    visual_presence: Literal["on_screen", "voice_only", "mentioned_only"] = "on_screen"
    role: str | None = None
    gender: str | None = None
    age: str | None = None
    appearance: str | None = None
    core_requirement: str | None = None
    personality: str | None = None
    background: str | None = None
    has_dialogue: bool = False
    dialogue_evidence: list[str] = Field(default_factory=list)
    context_codes: list[str] = Field(default_factory=list)


class SceneAttributesV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    context_code: str | None = None
    render_mode: Literal["animated", "live_action"] | None = None
    scene_no: str | None = None
    interior_exterior: str | None = None
    time: str | None = None
    location: str | None = None
    atmosphere: str | None = None
    narrative_function: str | None = None
    spatial_zones: list[str] = Field(default_factory=list)
    visual_goal: str | None = None
    key_props: list[str] = Field(default_factory=list)
    key_prop_codes: list[str] = Field(default_factory=list)
    continuity_risks: list[str] = Field(default_factory=list)
    episode_code: str | None = None


class PropAttributesV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    level: str | None = None
    prop_type: str | None = None
    appearance: str | None = None
    source: str | None = None
    status_change: str | None = None
    variant_states: list[str] = Field(default_factory=list)
    usage_scene: str | None = None
    scene_description: str | None = None
    context_codes: list[str] = Field(default_factory=list)


class AssetLinkV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_asset_key: str = Field(min_length=1)
    target_asset_code: str = Field(min_length=1)
    target_display_code: str = Field(min_length=1)
    target_display_label: str = Field(min_length=1)


class RelationEvidenceV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_code: str | None = None
    storyboard_code: str | None = None
    script_segment_code: str | None = None
    source_field: str | None = None
    excerpt: str | None = None


class AssetRelationsV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    relationship_text: str | None = None
    scene_links: list[AssetLinkV1] = Field(default_factory=list)
    owner_character: AssetLinkV1 | None = None
    storyboard_codes: list[str] = Field(default_factory=list)
    script_segment_codes: list[str] = Field(default_factory=list)
    evidence: list[RelationEvidenceV1] = Field(default_factory=list)
    confidence: Literal["high", "medium", "low", "unknown"] = "unknown"


class AssetReviewItemV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    review_id: str = Field(min_length=1)
    asset_key: str = Field(min_length=1)
    target_field: str = Field(min_length=1)
    issue: str = Field(min_length=1)
    suggestion: str | None = None
    status: str = "pending"
    severity: Literal["warning", "error"] = "warning"
    requirement_level: Literal["required", "optional"] = "required"
    source: str = "agent"
    source_path: str | None = None
    human_input: str | None = None


class GlobalAssetReviewItemV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    review_id: str = Field(min_length=1)
    target_scope: Literal["inventory"] = "inventory"
    target_field: str = Field(min_length=1)
    issue: str = Field(min_length=1)
    suggestion: str | None = None
    status: str = "pending"
    severity: Literal["warning", "error"] = "warning"
    requirement_level: Literal["required", "optional"] = "required"
    source: str = "agent"
    source_path: str | None = None
    human_input: str | None = None


class AssetSourceV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    source_kind: Literal["agent", "human", "confirmed_inventory"] = "agent"
    agent_run_id: str | None = None
    skill_name: str | None = None
    raw_group: Literal["assets", "characters", "scenes", "props"]
    raw_index: int = Field(ge=0)


class NormalizationRepairV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    field: str
    action: str
    before: Any = None
    after: Any = None
    reason: str


AssetAttributesV1 = CharacterAttributesV1 | SceneAttributesV1 | PropAttributesV1


class AssetNormalizationV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    normalization_version: NormalizationVersion = "AssetNormalization.v1"
    client_asset_key: str = Field(min_length=1)
    asset_code: str = Field(min_length=1)
    display_code: str = Field(min_length=1)
    display_label: str = Field(min_length=1)
    code_state: AssetCodeStateV1 = "candidate"
    asset_type: AssetTypeV1
    name: str = Field(min_length=1)
    description: str | None = None
    priority: AssetPriorityV1
    attributes: AssetAttributesV1
    relations: AssetRelationsV1 = Field(default_factory=AssetRelationsV1)
    review_items: list[AssetReviewItemV1] = Field(default_factory=list)
    source: AssetSourceV1
    repairs: list[NormalizationRepairV1] = Field(default_factory=list)

    @model_validator(mode="after")
    def attributes_match_asset_type(self) -> "AssetNormalizationV1":
        expected = {
            "character": CharacterAttributesV1,
            "scene": SceneAttributesV1,
            "prop": PropAttributesV1,
        }[self.asset_type]
        if not isinstance(self.attributes, expected):
            raise ValueError(f"{self.asset_type} requires {expected.__name__}")
        expected_display_prefix = {"character": "R", "scene": "SC", "prop": "P"}[self.asset_type]
        if not self.display_code.startswith(expected_display_prefix):
            raise ValueError(f"{self.asset_type} has an invalid display_code")
        if not self.asset_code.endswith(f"-{self.display_code}"):
            raise ValueError("asset_code and display_code must identify the same asset")
        if self.display_label != f"{self.display_code}-{self.name}":
            raise ValueError("display_label must be derived from display_code and name")
        if any(item.asset_key != self.client_asset_key for item in self.review_items):
            raise ValueError("review_items must belong to their containing asset")
        return self


class AssetNormalizationContextV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_prefix: str
    source_kind: Literal["agent", "human", "confirmed_inventory"] = "agent"
    agent_run_id: str | None = None
    skill_name: str | None = "asset-extract"


class AssetNormalizationResultV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    normalization_version: NormalizationVersion = "AssetNormalization.v1"
    assets: list[AssetNormalizationV1] = Field(default_factory=list)
    global_review_items: list[GlobalAssetReviewItemV1] = Field(default_factory=list)

    @model_validator(mode="after")
    def inventory_is_unique_and_closed(self) -> "AssetNormalizationResultV1":
        for field in ("client_asset_key", "asset_code", "display_code"):
            values = [getattr(asset, field) for asset in self.assets]
            if len(values) != len(set(values)):
                raise ValueError(f"assets must have unique {field}")
        by_key = {asset.client_asset_key: asset for asset in self.assets}
        review_keys: set[tuple[str, str]] = set()
        for asset in self.assets:
            for item in asset.review_items:
                review_key = (item.asset_key, item.review_id)
                if review_key in review_keys:
                    raise ValueError("review_id must be unique within an asset")
                review_keys.add(review_key)
            for link in asset.relations.scene_links:
                self._validate_link(link, by_key, expected_type="scene")
            if asset.relations.owner_character is not None:
                self._validate_link(asset.relations.owner_character, by_key, expected_type="character")
        global_review_ids = [item.review_id for item in self.global_review_items]
        if len(global_review_ids) != len(set(global_review_ids)):
            raise ValueError("global review_id must be unique")
        return self

    @staticmethod
    def _validate_link(
        link: AssetLinkV1,
        by_key: dict[str, AssetNormalizationV1],
        *,
        expected_type: AssetTypeV1,
    ) -> None:
        target = by_key.get(link.target_asset_key)
        if target is None:
            raise ValueError("asset relation target must exist in the same inventory")
        if target.asset_type != expected_type:
            raise ValueError(f"asset relation target must be a {expected_type}")
        if (
            link.target_asset_code != target.asset_code
            or link.target_display_code != target.display_code
            or link.target_display_label != target.display_label
        ):
            raise ValueError("asset relation target identity does not match its inventory asset")
